File: app/services/retention.py
from app import db, gmail_client


def cleanup_retention(account: dict, service) -> None:
    """Trash emails that exceed per-label or global retention rules."""
    account_id = account["id"]
    email_addr = account["email"]
    try:
        retention = db.get_retention(account_id)
        trashed_ids = set()

        exempt_names = {e["label_name"] for e in retention.get("exemptions", [])}
        exempt_lower = {n.lower() for n in exempt_names}

        for rule in retention["labels"]:
            if rule["label_name"].lower() in exempt_lower:
                continue
            ids = gmail_client.fetch_emails_older_than(service, rule["days"], rule["label_name"])
            new_ids = [i for i in ids if i not in trashed_ids]
            if new_ids:
                gmail_client.batch_trash_emails(service, new_ids)
                trashed_ids.update(new_ids)
                db.add_log(
                    "INFO",
                    f"[{email_addr}] Retention: trashed {len(new_ids)} email(s) with label "
                    f"'{rule['label_name']}' older than {rule['days']} day(s).",
                )

        if retention["global_days"]:
            excluded = list({rule["label_name"] for rule in retention["labels"]} | exempt_names)
            ids = gmail_client.fetch_emails_older_than(service, retention["global_days"], excluded_labels=excluded)
            new_ids = [i for i in ids if i not in trashed_ids]
            if new_ids:
                gmail_client.batch_trash_emails(service, new_ids)
                db.add_log(
                    "INFO",
                    f"[{email_addr}] Retention: trashed {len(new_ids)} email(s) older than "
                    f"{retention['global_days']} day(s) (global rule).",
                )
    except Exception as e:
        db.add_log("ERROR", f"[{email_addr}] Retention cleanup failed: {e}")

File: retention/retention.go
package retention

import (
	"context"
	"log/slog"

	"github.com/sloccy/ollamail/db"
	gmailpkg "github.com/sloccy/ollamail/gmail"
)

const maxRetentionIDs = 2500
const maxPages = 5

// Cleanup trashes emails that exceed retention rules for the given account.
func Cleanup(ctx context.Context, store *db.Store, svc *gmailpkg.Client, accountID int64) {
	defer func() {
		if r := recover(); r != nil {
			slog.Error("retention panic", "account_id", accountID, "err", r)
		}
	}()

	if err := cleanup(ctx, store, svc, accountID); err != nil {
		slog.Error("retention error", "account_id", accountID, "err", err)
	}
}

func cleanup(ctx context.Context, store *db.Store, svc *gmailpkg.Client, accountID int64) error {
	labelRules, err := store.GetLabelRetention(ctx, accountID)
	if err != nil {
		return err
	}

	exemptions, err := store.GetLabelExemptions(ctx, accountID)
	if err != nil {
		return err
	}
	exemptSet := make(map[string]bool, len(exemptions))
	for _, e := range exemptions {
		exemptSet[e.LabelName] = true
	}

	trashed := make(map[string]bool)

	// Per-label retention rules
	for _, rule := range labelRules {
		if exemptSet[rule.LabelName] {
			continue
		}
		for len(trashed) < maxRetentionIDs {
			ids, err := gmailpkg.FetchEmailsOlderThan(ctx, svc, int(rule.Days), rule.LabelName, nil, maxPages)
			if err != nil {
				slog.Error("fetch older emails", "label", rule.LabelName, "err", err)
				break
			}
			var toTrash []string
			for _, id := range ids {
				if !trashed[id] {
					toTrash = append(toTrash, id)
					trashed[id] = true
				}
			}
			if len(toTrash) == 0 {
				break
			}
			if err := gmailpkg.BatchTrashEmails(ctx, svc, toTrash); err != nil {
				slog.Error("trash emails", "label", rule.LabelName, "err", err)
			}
			if len(ids) < 500 {
				break // no more pages
			}
		}
	}

	// Global retention rule
	retention, err := store.GetAccountRetention(ctx, accountID)
	if err != nil {
		return nil //nolint:nilerr // no global rule configured
	}
	if !retention.GlobalDays.Valid {
		return nil
	}

	// Build exclusion list: labels with specific rules + exemptions
	var excludeLabels []string
	for _, rule := range labelRules {
		excludeLabels = append(excludeLabels, rule.LabelName)
	}
	for _, e := range exemptions {
		excludeLabels = append(excludeLabels, e.LabelName)
	}

	for len(trashed) < maxRetentionIDs {
		ids, err := gmailpkg.FetchEmailsOlderThan(ctx, svc, int(retention.GlobalDays.Int64), "", excludeLabels, maxPages)
		if err != nil {
			slog.Error("fetch older emails (global)", "err", err)
			break
		}
		var toTrash []string
		for _, id := range ids {
			if !trashed[id] {
				toTrash = append(toTrash, id)
				trashed[id] = true
			}
		}
		if len(toTrash) == 0 {
			break
		}
		if err := gmailpkg.BatchTrashEmails(ctx, svc, toTrash); err != nil {
			slog.Error("trash emails (global)", "err", err)
		}
		if len(ids) < 500 {
			break
		}
	}
	return nil
}
